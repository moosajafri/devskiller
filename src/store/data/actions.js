export const LOAD_EMPLOYEES = "LOAD_EMPLOYEES"
export const LOAD_PROJECTS = "LOAD_PROJECTS"

export const loadEmployees = ({ employees }) => ({
  type: LOAD_EMPLOYEES,
  employees,
})

export const loadProjects = ({ projects }) => ({
  type: LOAD_PROJECTS,
  projects,
})

export const actions = {
  loadEmployees,
  loadProjects,
}
