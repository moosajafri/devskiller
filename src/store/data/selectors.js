export const someSelector = () => {
  // to be implemented...
}
