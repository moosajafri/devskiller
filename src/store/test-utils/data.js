export const dataSet = {
  projects: () => require('./projects.json'),
  employees: () => require('./employees.json'),
}
