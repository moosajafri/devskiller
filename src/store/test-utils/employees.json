[
  {
    "id": 302356259,
    "salary": 9577,
    "firstName": "Jackson",
    "lastName": "Murray",
    "skills": [
      "technical support",
      "customer support"
    ],
    "contract": "permanent",
    "position": "Software Engineer"
  },
  {
    "id": 891265620,
    "salary": 5969,
    "firstName": "Semih",
    "lastName": "Agostini",
    "skills": [
      "C#",
      ".net",
      "EntityFramework",
      "VisualStudio",
      "VSTS",
      "AngularJS",
      "RxJS",
      "jQuery",
      "CSS",
      "HTML",
      "algorithms",
      "scrum",
      "testing"
    ],
    "contract": "contract",
    "position": "Senior DevOps Engineer"
  },
  {
    "id": 85266,
    "salary": 4377,
    "firstName": "Kelli",
    "lastName": "Stamm",
    "skills": [
      "C#",
      ".net",
      "EntityFramework",
      "OData",
      "IIS",
      "VisualStudio",
      "VSTS",
      "JavaScript",
      "Angular",
      "jQuery",
      "CSS",
      "HTML",
      "DDD",
      "SOA",
      "MVC",
      "Scalability",
      "algorithms",
      "scrum",
      "testing"
    ],
    "contract": "permanent",
    "position": ".Net Engineer"
  },
  {
    "id": 8184090,
    "salary": 3288,
    "firstName": "Maëlys",
    "lastName": "Roy",
    "skills": [
      "Java",
      "scala",
      "JVM",
      "spring",
      "JSP",
      "C#",
      ".net",
      "EntityFramework",
      "OData",
      "IIS",
      "Rx.net",
      "VisualStudio",
      "data structures",
      "scrum",
      "testing"
    ],
    "contract": "permanent",
    "position": "Senior Cloud Developer"
  },
  {
    "id": 9432126,
    "salary": 5998,
    "firstName": "Thalea",
    "lastName": "Heberstreit",
    "skills": [
      "SOA",
      "MVC",
      "Scalability",
      "Security",
      "C#",
      ".net",
      "EntityFramework",
      "OData",
      "IIS",
      "VisualStudio",
      "data structures",
      "scrum"
    ],
    "contract": "contract",
    "position": "Senior Cloud Engineer"
  },
  {
    "id": 528605,
    "salary": 1249,
    "firstName": "Eva",
    "lastName": "Janssen",
    "skills": [
      "C#",
      ".net",
      "EntityFramework",
      "OData",
      "IIS",
      "Rx.net",
      "VisualStudio",
      "VSTS",
      "react",
      "redux",
      "AngularJS",
      "jQuery",
      "CSS",
      "HTML",
      "SOA",
      "MVC",
      "data structures",
      "algorithms",
      "scrum",
      "testing",
      "management"
    ],
    "contract": "permanent",
    "position": ".Net Developer"
  },
  {
    "id": 51767706,
    "salary": 5521,
    "firstName": "Chance",
    "lastName": "Johns",
    "skills": [
      "JavaScript",
      "TypeScript",
      "react",
      "jQuery",
      "HTML"
    ],
    "contract": "permanent",
    "position": ".Net Developer"
  },
  {
    "id": 9928851,
    "salary": 9213,
    "firstName": "Teseo",
    "lastName": "Martini",
    "skills": [
      "MVC",
      "Scalability",
      "Java",
      "JVM",
      "spring",
      "data structures",
      "scrum",
      "testing",
      "management"
    ],
    "contract": "contract",
    "position": "Java Developer"
  },
  {
    "id": 85836898,
    "salary": 5289,
    "firstName": "Tacjana",
    "lastName": "Łukasiewicz",
    "skills": [
      "Java",
      "scala",
      "JVM",
      "spring",
      "JSP",
      "Hibernate",
      "sql"
    ],
    "contract": "permanent",
    "position": "DevOps Engineer"
  },
  {
    "id": 227947241,
    "salary": 7512,
    "firstName": "Josiane",
    "lastName": "Rodriguez",
    "skills": [
      "MVC",
      "Scalability",
      "C#",
      ".net",
      "EntityFramework",
      "OData",
      "IIS",
      "VisualStudio",
      "VSTS",
      "data structures",
      "algorithms",
      "scrum",
      "testing"
    ],
    "contract": "contract",
    "position": "Software Engineer"
  },
  {
    "id": 1791071,
    "salary": 6579,
    "firstName": "Sophie",
    "lastName": "Klein",
    "skills": [
      "Java",
      "spring",
      "JSP",
      "JavaScript",
      "TypeScript",
      "AngularJS",
      "jQuery",
      "CSS",
      "HTML",
      "data structures",
      "algorithms",
      "scrum",
      "testing"
    ],
    "contract": "permanent",
    "position": "Senior Cloud Developer"
  },
  {
    "id": 99835525,
    "salary": 3147,
    "firstName": "Lou",
    "lastName": "Rousseau",
    "skills": [
      "C#",
      ".net",
      "EntityFramework",
      "IIS",
      "Rx.net",
      "VisualStudio",
      "VSTS",
      "JavaScript",
      "redux",
      "jQuery",
      "CSS",
      "HTML",
      "data structures",
      "scrum",
      "testing"
    ],
    "contract": "contract",
    "position": "Frontend Engineer"
  },
  {
    "id": 614789748,
    "salary": 4796,
    "firstName": "Adelajda",
    "lastName": "Stachowiak",
    "skills": [
      "C#",
      ".net",
      "EntityFramework",
      "IIS",
      "Rx.net",
      "VisualStudio",
      "VSTS",
      "JavaScript",
      "jQuery",
      "CSS",
      "HTML"
    ],
    "contract": "contract",
    "position": "Junior .Net Developer"
  },
  {
    "id": 382582,
    "salary": 5206,
    "firstName": "Fatima",
    "lastName": "Rossetti",
    "skills": [
      "C#",
      ".net",
      "EntityFramework",
      "Rx.net",
      "VisualStudio",
      "VSTS",
      "JavaScript",
      "AngularJS",
      "jQuery",
      "CSS",
      "HTML"
    ],
    "contract": "permanent",
    "position": ".Net Engineer"
  },
  {
    "id": 168575215,
    "salary": 3267,
    "firstName": "Berta",
    "lastName": "Negrete",
    "skills": [
      "MVC",
      "Scalability",
      "C#",
      ".net",
      "EntityFramework",
      "IIS",
      "Rx.net",
      "VisualStudio",
      "VSTS",
      "data structures",
      "algorithms",
      "scrum",
      "testing"
    ],
    "contract": "contract",
    "position": "Junior Cloud Developer"
  },
  {
    "id": 64801242,
    "salary": 3883,
    "firstName": "Soledad",
    "lastName": "Agosto",
    "skills": [
      "C#",
      ".net",
      "EntityFramework",
      "OData",
      "IIS",
      "VisualStudio",
      "VSTS",
      "JavaScript",
      "react",
      "jQuery",
      "CSS",
      "HTML",
      "SOA",
      "MVC",
      "Scalability",
      "algorithms",
      "scrum",
      "testing"
    ],
    "contract": "permanent",
    "position": "Junior .Net Engineer"
  }
]