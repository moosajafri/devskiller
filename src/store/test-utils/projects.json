[
  {
    "id": "01e226fa-9c4a-48ad-8aab-5dd20d2dc1ba",
    "name": "Fantastic Concrete Pants",
    "budget": 744000,
    "team": [
      {
        "id": 85266,
        "name": "Kelli Stamm"
      }
    ],
    "manager": 528605
  },
  {
    "id": "627e647e-21ec-4b1a-b8f6-27a3cde4c7d1",
    "name": "Handcrafted Wooden Computer",
    "budget": 484000,
    "team": [
      {
        "id": 168575215,
        "name": "Berta Negrete"
      }
    ],
    "manager": 528605
  },
  {
    "id": "21323ef7-5cf9-4754-b7f4-5aa2b3301d54",
    "name": "Handmade Fresh Chair",
    "budget": 5124000,
    "team": [
      {
        "id": 891265620,
        "name": "Semih Agostini"
      },
      {
        "id": 85266,
        "name": "Kelli Stamm"
      },
      {
        "id": 168575215,
        "name": "Berta Negrete"
      }
    ],
    "manager": 528605
  },
  {
    "id": "92829075-5477-48e8-a69b-b29dff45abcc",
    "name": "Practical Concrete Shirt",
    "budget": 7430000,
    "team": [
      {
        "id": 9432126,
        "name": "Thalea Heberstreit"
      },
      {
        "id": 85836898,
        "name": "Tacjana Łukasiewicz"
      }
    ],
    "manager": 528605
  },
  {
    "id": "aff9a279-77a5-453e-b413-4bf2e2f9723c",
    "name": "Intelligent Rubber Soap",
    "budget": 3782000,
    "team": [
      {
        "id": 8184090,
        "name": "Maëlys Roy"
      },
      {
        "id": 9432126,
        "name": "Thalea Heberstreit"
      }
    ],
    "manager": 9928851
  },
  {
    "id": "df3f32b4-112e-4f79-bd62-63cc09151369",
    "name": "Gorgeous Rubber Fish",
    "budget": 3605000,
    "team": [
      {
        "id": 8184090,
        "name": "Maëlys Roy"
      },
      {
        "id": 51767706,
        "name": "Chance Johns"
      },
      {
        "id": 1791071,
        "name": "Sophie Klein"
      }
    ],
    "manager": 9928851
  }
]