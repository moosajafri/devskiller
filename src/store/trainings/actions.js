export const CREATE_TRAINING = "CREATE_TRAINING"
export const UPDATE_TRAINING = "UPDATE_TRAINING"
export const CLEAR_PARTICIPANTS = "CLEAR_PARTICIPANTS"
export const ENROLL_EMPLOYEE_TO_TRAINING = "ENROLL_EMPLOYEE_TO_TRAINING"
export const REMOVE_EMPLOYEE_FROM_TRAINING = "REMOVE_EMPLOYEE_FROM_TRAINING"

export const createTraining = ({ name, trainingSkills, level }) => ({
  type: CREATE_TRAINING,
  name, trainingSkills, level
})

export const updateTraining = ({ id, name, trainingSkills, level }) => ({
  type: UPDATE_TRAINING,
  id, name, trainingSkills, level
})

export const clearParticipants = ({ trainingID }) => ({
  type: CLEAR_PARTICIPANTS,
  trainingID,
})

export const enrollEmployeeToTraining = ({ employeeID, trainingID }) => ({
  type: ENROLL_EMPLOYEE_TO_TRAINING,
  employeeID,
  trainingID
})

export const removeEmployeeFromTraining = ({ employeeID, trainingID }) => ({
  type: REMOVE_EMPLOYEE_FROM_TRAINING,
  employeeID,
  trainingID
})

export const actions = {
  createTraining,
  updateTraining,
  clearParticipants,
  enrollEmployeeToTraining,
  removeEmployeeFromTraining,
}
