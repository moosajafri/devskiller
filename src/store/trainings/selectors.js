export const getGeneralTrainingsInfo = (state) => {
  // implementation here
}

export const getTrainingParticipantsInfo = (state, trainingID) => {
  // implementation here
}

export const getTrainingSkillsInfo = (state, trainingID) => {
  // implementation here
}
