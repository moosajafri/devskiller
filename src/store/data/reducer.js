import produce from "immer";

import { actions } from './actions';

const initialState = {
  employees: [],
  projects: [],
}

export const dataReducer = (state = initialState, action) => {
  switch(action.type){
    case "LOAD_EMPLOYEES":
      return {
        ...state,
        employees: action.employees,
      }

    case "LOAD_PROJECTS":
      return {
        ...state,
        projects: action.projects,
      }

    default:
      return state
  }
}
