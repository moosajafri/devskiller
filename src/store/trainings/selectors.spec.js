import { createTraining, enrollEmployeeToTraining, clearParticipants, removeEmployeeFromTraining } from "./actions";
import { getTrainingSkillsInfo, getTrainingParticipantsInfo, getGeneralTrainingsInfo } from './selectors';
import { loadEmployees } from "../data";

import { initStoreWithActions, dataSet } from "../test-utils";

const noParticipants = {
  name: "#A No Participants",
  actions: [
    loadEmployees({ employees: dataSet.employees() }),
    createTraining({
      name: "From React & Redux to React Hooks",
      trainingSkills: ['react', 'redux', 'react-hooks'],
      level: "ADVANCED",
    }),
  ]
}

const someParticipantsEnrolled = {
  name: "#B Some Participants Enrolled",
  actions: [
    loadEmployees({ employees: dataSet.employees() }),
    createTraining({
      name: "From React & Redux to React Hooks",
      trainingSkills: ['react', 'redux', 'react-hooks'],
      level: "ADVANCED",
    }),
    enrollEmployeeToTraining({ employeeID: 51767706, trainingID: 1 }),
    enrollEmployeeToTraining({ employeeID: 64801242, trainingID: 1 }),
    enrollEmployeeToTraining({ employeeID: 528605, trainingID: 1 }),
    enrollEmployeeToTraining({ employeeID: 302356259, trainingID: 1 }),
    enrollEmployeeToTraining({ employeeID: 8184090, trainingID: 1 }),
  ]
}

const someParticipantsModified = {
  name: "#C Some Participants Modified",
  actions: [
    loadEmployees({ employees: dataSet.employees() }),
    createTraining({
      name: "From React & Redux to React Hooks",
      trainingSkills: ['react', 'redux', 'react-hooks'],
      level: "ADVANCED",
    }),
    createTraining({
      name: "Angular for AngularJS developers",
      trainingSkills: ['Angular', 'JavaScript', 'TypeScript', 'RxJS'],
      level: "BEGINNER",
    }),
    enrollEmployeeToTraining({ employeeID: 168575215, trainingID: 2 }),
    enrollEmployeeToTraining({ employeeID: 51767706, trainingID: 1 }),
    clearParticipants({ trainingID: 1 }),
    enrollEmployeeToTraining({ employeeID: 382582, trainingID: 2 }),
    enrollEmployeeToTraining({ employeeID: 1791071, trainingID: 1 }),
    enrollEmployeeToTraining({ employeeID: 528605, trainingID: 1 }),
    removeEmployeeFromTraining({ employeeID: 1791071, trainingID: 1 }),
  ]
}

describe('Trainings Selectors', () => {

  describe('getGeneralTrainingsInfo', () => {
    const testcases = [{
      ...noParticipants,
    }, {
      ...someParticipantsEnrolled,
    }, {
      ...someParticipantsModified,
    }];

    testcases.forEach(({ name, actions }) =>
      it(`should generate general training info for testcase:\n"${name}"`, () => {
        const store = initStoreWithActions(actions);

        const result = getGeneralTrainingsInfo(store.getState());
        expect(result).toMatchSnapshot();
      })
    );
  });

  describe('getTrainingParticipantsInfo', () => {
    const testcases = [{
      ...noParticipants,
      trainingID: 1,
    }, {
      ...someParticipantsEnrolled,
      trainingID: 1,
    }, {
      ...someParticipantsModified,
      trainingID: 1,
    }];

    testcases.forEach(({ name, actions, trainingID }) => {
      it(`should generate training participants info for testcase:\n"${name} (trainingID: ${trainingID})"`, () => {
        const store = initStoreWithActions(actions);

        const result = getTrainingParticipantsInfo(store.getState(), trainingID);
        expect(result).toMatchSnapshot();
      })
    });
  });

  describe('getTrainingSkillsInfo', () => {
    const testcases = [{
      ...noParticipants,
      trainingID: 1,
    }, {
      ...someParticipantsEnrolled,
      trainingID: 1,
    }, {
      ...someParticipantsModified,
      trainingID: 1,
    }];

    testcases.forEach(({ name, actions, trainingID }) => {
      it(`should generate skills info for testcase:\n"${name} (trainingID: ${trainingID})"`, () => {
        const store = initStoreWithActions(actions);

        const result = getTrainingSkillsInfo(store.getState(), trainingID);
        expect(result).toMatchSnapshot();
      })
    });
  });

});
