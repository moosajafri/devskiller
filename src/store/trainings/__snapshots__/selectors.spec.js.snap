// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trainings Selectors getGeneralTrainingsInfo should generate general training info for testcase:
"#A No Participants" 1`] = `
Array [
  Object {
    "id": 1,
    "level": "ADVANCED",
    "name": "From React & Redux to React Hooks",
    "participants": 0,
  },
]
`;

exports[`Trainings Selectors getGeneralTrainingsInfo should generate general training info for testcase:
"#B Some Participants Enrolled" 1`] = `
Array [
  Object {
    "id": 1,
    "level": "ADVANCED",
    "name": "From React & Redux to React Hooks",
    "participants": 5,
  },
]
`;

exports[`Trainings Selectors getGeneralTrainingsInfo should generate general training info for testcase:
"#C Some Participants Modified" 1`] = `
Array [
  Object {
    "id": 2,
    "level": "BEGINNER",
    "name": "Angular for AngularJS developers",
    "participants": 2,
  },
  Object {
    "id": 1,
    "level": "ADVANCED",
    "name": "From React & Redux to React Hooks",
    "participants": 1,
  },
]
`;

exports[`Trainings Selectors getTrainingParticipantsInfo should generate training participants info for testcase:
"#A No Participants (trainingID: 1)" 1`] = `Array []`;

exports[`Trainings Selectors getTrainingParticipantsInfo should generate training participants info for testcase:
"#B Some Participants Enrolled (trainingID: 1)" 1`] = `
Array [
  Object {
    "id": 64801242,
    "name": "Soledad Agosto",
    "position": "Junior .Net Engineer",
  },
  Object {
    "id": 528605,
    "name": "Eva Janssen",
    "position": ".Net Developer",
  },
  Object {
    "id": 51767706,
    "name": "Chance Johns",
    "position": ".Net Developer",
  },
  Object {
    "id": 302356259,
    "name": "Jackson Murray",
    "position": "Software Engineer",
  },
  Object {
    "id": 8184090,
    "name": "MaÃ«lys Roy",
    "position": "Senior Cloud Developer",
  },
]
`;

exports[`Trainings Selectors getTrainingParticipantsInfo should generate training participants info for testcase:
"#C Some Participants Modified (trainingID: 1)" 1`] = `
Array [
  Object {
    "id": 528605,
    "name": "Eva Janssen",
    "position": ".Net Developer",
  },
]
`;

exports[`Trainings Selectors getTrainingSkillsInfo should generate skills info for testcase:
"#A No Participants (trainingID: 1)" 1`] = `
Object {
  "react": Object {
    "have": 0,
    "not-have": 0,
  },
  "react-hooks": Object {
    "have": 0,
    "not-have": 0,
  },
  "redux": Object {
    "have": 0,
    "not-have": 0,
  },
}
`;

exports[`Trainings Selectors getTrainingSkillsInfo should generate skills info for testcase:
"#B Some Participants Enrolled (trainingID: 1)" 1`] = `
Object {
  "react": Object {
    "have": 3,
    "not-have": 2,
  },
  "react-hooks": Object {
    "have": 0,
    "not-have": 5,
  },
  "redux": Object {
    "have": 1,
    "not-have": 4,
  },
}
`;

exports[`Trainings Selectors getTrainingSkillsInfo should generate skills info for testcase:
"#C Some Participants Modified (trainingID: 1)" 1`] = `
Object {
  "react": Object {
    "have": 1,
    "not-have": 0,
  },
  "react-hooks": Object {
    "have": 0,
    "not-have": 1,
  },
  "redux": Object {
    "have": 1,
    "not-have": 0,
  },
}
`;
